Comply with the received Record Size Limit extension

KloolK opened the pull request:
Fixes #7010

## Gatekeeper checklist

- [ ] **changelog** no, feature is not complete for production use
- [ ] **backport** not required
- [x] **tests** provided

KloolK commented:
@yanesca I have two open questions regarding my implementation and would like to get some feedback on these:

1:
This part https://github.com/Mbed-TLS/mbedtls/pull/7455/files#diff-f60739d09b65e7b43187688f40e19d5c9825972f8690c0f6f9611431f2cf738eR1692-R1698 is currently untested.
I did some manual tests and it seems fine. But there are no automated tests at the moment, since I don't see a way to trigger this codepath with the GnuTLS client.
Is that fine for now?

2:
Regarding what I describe here: https://github.com/Mbed-TLS/mbedtls/pull/7455/files#diff-f60739d09b65e7b43187688f40e19d5c9825972f8690c0f6f9611431f2cf738eR1700-R1714.
Where do you think is the correct place to perform this subtraction? Ideally some place where the total application data length that is currently attempted to be written via `ssl_write_real()` (`len`) is available (to be able to consider the actual padding, which depends on the data length).

Thanks!

KloolK commented on tests/scripts/all.sh:
```diff
@@ -2953,10 +2962,10 @@ component_build_psa_accel_alg_hkdf() {
     scripts/config.py full
     scripts/config.py set MBEDTLS_PSA_CRYPTO_CONFIG
     scripts/config.py unset MBEDTLS_USE_PSA_CRYPTO
-    scripts/config.py unset MBEDTLS_SSL_PROTO_TLS1_3
```
Removing duplicate line (already exists 3 lines below)

KloolK commented:
I removed `MBEDTLS_SSL_RECORD_SIZE_LIMIT` from the exclude list in `config.py`: https://github.com/Mbed-TLS/mbedtls/pull/7455/files#diff-e9ffa9c36082d5a0b2ae4098b1b77d020a8d57d159bc89255fcad17287bc3ed9L218 since enabling it should not break unrelated tests anymore.

This made the changes in `all.sh` necessary: https://github.com/Mbed-TLS/mbedtls/pull/7455/files#diff-e683396c93d3af1ea2db5a3d94788da3916ed084f3eb8be5e5f7c0e429badf39

However, `all_u16-test_depends_py_hashes` and `all_u16-test_depends_py_hashes_psa` still fail, due to `MBEDTLS_SSL_RECORD_SIZE_LIMIT` being enabled without `MBEDTLS_SSL_PROTO_TLS1_3`.

Now I'm wondering how to fix this properly. Any hints?

yanesca commented:
>I did some manual tests and it seems fine. But there are no automated tests at the moment, since I don't see a way to trigger this codepath with the GnuTLS client.
Is that fine for now?

Yes, it is. For now we just want to do some smoke testing. Adding unit tests and extensive testing is out of scope for this PR and will be done in https://github.com/Mbed-TLS/mbedtls/issues/7027.

> Where do you think is the correct place to perform this subtraction?

We can calculate it right here rounding down to a multiple of `MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY`. (Messages will be padded to a multiple of `MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY`, anything longer than that would be padded to something longer than the record size limit.)

yanesca commented:
> Now I'm wondering how to fix this properly. Any hints?

I suggest we don't remove MBEDTLS_SSL_RECORD_SIZE_LIMIT from the exclude list in config.py just yet. We can do that later after the feature is complete and we have added extensive unit tests.

KloolK commented:
> We can calculate it right here rounding down to a multiple of `MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY`. (Messages will be padded to a multiple of `MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY`, anything longer than that would be padded to something longer than the record size limit.)

Done, with subtracting 1 byte after rounding to account for the content type that will be added afterwards.

> I suggest we don't remove MBEDTLS_SSL_RECORD_SIZE_LIMIT from the exclude list in config.py just yet. We can do that later after the feature is complete and we have added extensive unit tests.

Reverted both the change to `config.py` and the related changes to `all.sh`.

yanesca commented:
@KloolK This PR is something that we still need and a contribution we want to take, we are just struggling to find the review bandwidth for it at the moment. I just wanted to give you a heads up that it looks like it will take some time until we will be able to schedule the review. Thanks again for your contribution and thank you for your patience.

xkqian commented on include/mbedtls/ssl.h:
```diff
@@ -1184,6 +1184,11 @@ struct mbedtls_ssl_session {
     unsigned char MBEDTLS_PRIVATE(mfl_code);     /*!< MaxFragmentLength negotiated by peer */
 #endif /* MBEDTLS_SSL_MAX_FRAGMENT_LENGTH */
 
+/*!< Record Size Limit for outgoing data requested by peer */
```
```suggestion
/*!< RecordSizeLimit received by peer */
```

KloolK replied:
This makes it sounds as if the added field is the limit the peer is going to use for data sent to us.

waleed-elmelegy-arm replied:
Done.

xkqian commented on library/ssl_misc.h:
```diff
@@ -450,6 +450,24 @@ size_t mbedtls_ssl_get_output_max_frag_len(const mbedtls_ssl_context *ssl);
 
+#if defined(MBEDTLS_SSL_RECORD_SIZE_LIMIT)
+/**
+ * \brief          Return the record size limit (in bytes) for
+ *                 the output buffer. This is less than the value requested by the
+ *                 peer (using RFC 8449), since it subtracts the space required for the
```
```suggestion
 * \brief          Return the RecordSizeLimit  (in bytes) for
 *                 the output buffer. This is less than the value requested by the
 *                 peer (See RFC 8449), since it subtracts the space required for the
```

waleed-elmelegy-arm replied:
Done.

xkqian commented on library/ssl_misc.h:
```diff
@@ -450,6 +450,24 @@ size_t mbedtls_ssl_get_output_max_frag_len(const mbedtls_ssl_context *ssl);
+ *
+ * \return         Current record size limit for the output buffer.
+ */
+size_t mbedtls_ssl_get_output_record_size_limit(const mbedtls_ssl_context *ssl);
```
Why here return size_t but the filed defined in mbedtls_ssl_session{} is uint16_t?

KloolK replied:
I wanted to make the API similar to `mbedtls_ssl_get_input/output_max_frag_len()`. I don't for sure why `size_t` was chosen there, but I would guess to be prepared for future TLS versions that might allow record sizes larger than `uint16_t`.
I'd prefer keeping it a `uint16_t` in the session in order to not make it unnecessarily large.

xkqian commented on library/ssl_tls13_generic.c:
```diff
@@ -1673,13 +1673,47 @@ int mbedtls_ssl_tls13_parse_record_size_limit_ext(mbedtls_ssl_context *ssl,
+        record_size_limit = ssl_compute_internal_record_size_limit(ssl->session->record_size_limit);
+    }
+
+    // TODO: this is currently untested
```
Maybe you can make use of the seesion resumption cases(The cases titled with: TLS 1.3: NewSessionTicket:...) to test it.

xkqian commented on tests/ssl-opt.sh:
```diff
@@ -4733,8 +4735,162 @@ run_test    "Record Size Limit: TLS 1.3: Client-side parsing, debug output and f
-#            -s "Received alert \[110]: An unsupported extension was sent"
+#            -c "RecordSizeLimit: 16385 Bytes"
+
+# In the following (9) tests, --recordsize is the value used by the G_NEXT_CLI (3.7.2) to configure the
```
How about the test cases with the peer from OpenSSL.

waleed-elmelegy-arm replied:
OpenSSL does not support record size limit yet

tom-cosgrove-arm commented:
Hi @KloolK, can I check if you will have time to make the changes requested, or would you like us to finish off this PR?

tom-cosgrove-arm commented:
@waleed-elmelegy-arm

KloolK commented:
> Hi @KloolK, can I check if you will have time to make the changes requested, or would you like us to finish off this PR?

Hi @tom-cosgrove-arm I don't expect to have much time in the coming weeks unfortunately. Feel free to finish this off.

waleed-elmelegy-arm commented:
notes about this PR:
* In our interoperability tests with Mbedtls server and GnuTLS as a client our server sends a message with the certificate that exceeds the record size of 512 bytes, However this passes here because according to [this](https://gitlab.com/gnutls/gnutls/-/blob/master/lib/ext/record_size_limit.c?ref_type=heads#L90) the client treats the server sending the record size extension as an indication the record size extension sent before was accepted and we don't send the record size extension yet from the server, when we do in this [PR](https://github.com/Mbed-TLS/mbedtls/pull/8526) the test does fail.
* From the server side we don't check on the record size during the handshake only later when writing application data, during the handshake we rely on `MBEDTLS_SSL_OUT_CONTENT_LEN` which is defaulted to 16384, so this should be changed to use the record size, when I change this the 512 bytes test fail because certificate size is bigger than 512 bytes and since fragmentation is not supported there is nothing we can do.
So what I am proposing to do is add the proper checking to the server and remove the 512 bytes test unless there is a way to use a smaller certificate Let me know if anyone has any feedback on that.

yanesca commented:
I think we should be able to construct a SECP192 certificate that is shorter than 512 bytes.

tom-cosgrove-arm commented:
@waleed-elmelegy-arm There are still conflicts. I suggest rebasing against `development` rather than a merge, since we're early in the review cycle on this PR

waleed-elmelegy-arm commented on library/ssl_msg.c:
```diff
@@ -917,6 +917,7 @@ int mbedtls_ssl_encrypt_buf(mbedtls_ssl_context *ssl,
 #endif
     size_t add_data_len;
     size_t post_avail;
+    int max_out_record_len = mbedtls_ssl_get_max_out_record_payload(ssl);
```
I think the problem is checking on maximum fragment length according to this: https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L4262-L4264 we do not support currently maximum fragment length during handshake so I added a check for if we are in handshake or not.

yanesca replied:
Thank you @yanrayw, I think @waleed-elmelegy-arm 's solution should resolve the issue.

yanesca replied:
That doesn't add up, we fragment handshake messages in DTLS. How can the certificate be too big?

ronald-cron-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1188,6 +1188,11 @@ struct mbedtls_ssl_session {
     unsigned char MBEDTLS_PRIVATE(mfl_code);     /*!< MaxFragmentLength negotiated by peer */
 #endif /* MBEDTLS_SSL_MAX_FRAGMENT_LENGTH */
 
+/*!< RecordSizeLimit received by peer */
```
Otherwise, see comment line 1176 in this file, we miss some changes in test code: in ssl_populate_session() and ssl_serialize_session_save_load(). What about ssl_session_copy()?

ronald-cron-arm commented on tests/ssl-opt.sh:
```diff
@@ -4863,8 +4867,176 @@ run_test    "Record Size Limit: TLS 1.3: Client-side parsing, debug output and f
+# requires_config_enabled MBEDTLS_SSL_RECORD_SIZE_LIMIT
+# requires_config_enabled MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
+# run_test    "Record Size Limit: TLS 1.3: Server complies with record size limit (513), 1 fragment" \
+#             "$P_SRV debug_level=3 force_version=tls13 response_size=256" \
```
I thought the tests with a record size limit of 512 should fail. But looking at the CI (and locally for me) it does not seem the case. What am I missing?

ronald-cron-arm replied:
I just missed that they are commented out ...

waleed-elmelegy-arm replied:
The tests in this tocket will never fail becuase GnuTLS does not start to commit to the record size restrictions except when the server sends it's record size as well which GnuTLS treats as acknowledgment from the server to accepting the record size requested ( see https://gitlab.com/gnutls/gnutls/-/blob/master/lib/ext/record_size_limit.c?ref_type=heads#L90 ) so tests will start to fail in https://github.com/Mbed-TLS/mbedtls/pull/8526

ronald-cron-arm replied:
Ah that makes sense. Thanks.

ronald-cron-arm reviewed (changes requested):
I've reviewed the library code. Please find my comments below. I have started to look at the tests as well but not completely.

ronald-cron-arm commented on tests/ssl-opt.sh:
```diff
@@ -4863,8 +4867,176 @@ run_test    "Record Size Limit: TLS 1.3: Client-side parsing, debug output and f
+#             -s "RecordSizeLimit: 513 Bytes" \
+#             -s "ClientHello: record_size_limit(28) extension exists." \
+#             -s "Maximum outgoing record payload length is 511" \
+#             -s "1280 bytes written in 3 fragments"
```
For the record size limit of 513, I propose to do an handshake in PSK mode (no certificate), something like:
```
requires_gnutls_tls1_3                                                          
requires_gnutls_record_size_limit                                               
requires_all_configs_enabled MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C
requires_config_enabled MBEDTLS_SSL_RECORD_SIZE_LIMIT                           
requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED        
run_test    "Record Size Limit: TLS 1.3: Server complies with record size limit (513), 1 fragment" \
            "$P_SRV debug_level=3 force_version=tls13 tls13_kex_modes=psk \     
                    psk_list=Client_identity,6162636465666768696a6b6c6d6e6f70 \ 
                    response_size=256" \                                        
            "$G_NEXT_CLI localhost --priority=NORMAL:-VERS-ALL:+VERS-TLS1.3:+PSK --recordsize 512 \
                         --pskusername Client_identity --pskkey=6162636465666768696a6b6c6d6e6f70" \
            0 \                                                                 
            -s "RecordSizeLimit: 513 Bytes" \                                   
            -s "Maximum outgoing record payload length is 511" \                
            -s "256 bytes written in 1 fragments"                               
                                                                                
requires_gnutls_tls1_3                                                          
requires_gnutls_record_size_limit                                               
requires_all_configs_enabled MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C
requires_config_enabled MBEDTLS_SSL_RECORD_SIZE_LIMIT                           
requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED        
run_test    "Record Size Limit: TLS 1.3: Server complies with record size limit (513), 2 fragments" \
            "$P_SRV debug_level=3 force_version=tls13 tls13_kex_modes=psk \     
                    psk_list=Client_identity,6162636465666768696a6b6c6d6e6f70 \ 
                    response_size=768" \                                        
            "$G_NEXT_CLI localhost --priority=NORMAL:-VERS-ALL:+VERS-TLS1.3:+PSK --recordsize 512 \
                         --pskusername Client_identity --pskkey=6162636465666768696a6b6c6d6e6f70" \
            0 \                                                                 
            -s "RecordSizeLimit: 513 Bytes" \                                   
            -s "Maximum outgoing record payload length is 511" \                
            -s "768 bytes written in 2 fragments"                               
                                                                                
requires_gnutls_tls1_3                                                          
requires_gnutls_record_size_limit                                               
requires_all_configs_enabled MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C
requires_config_enabled MBEDTLS_SSL_RECORD_SIZE_LIMIT                           
requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED        
run_test    "Record Size Limit: TLS 1.3: Server complies with record size limit (513), 3 fragments" \
            "$P_SRV debug_level=3 force_version=tls13 tls13_kex_modes=psk \     
                    psk_list=Client_identity,6162636465666768696a6b6c6d6e6f70 \ 
                    response_size=1280" \                                       
            "$G_NEXT_CLI localhost --priority=NORMAL:-VERS-ALL:+VERS-TLS1.3:+PSK --recordsize 512 \
                         --pskusername Client_identity --pskkey=6162636465666768696a6b6c6d6e6f70" \
            0 \                                                                 
            -s "RecordSizeLimit: 513 Bytes" \                                   
            -s "Maximum outgoing record payload length is 511" \                
            -s "1280 bytes written in 3 fragments
```

ronald-cron-arm replied:
Otherwise, please align the test case dependencies of the following test cases with those of the three proposed ones.

ronald-cron-arm replied:
This comment is not addressed yet it seems. Thus for tests with certificate (sizes 1024, 2048 and 4096) the dependencies should rather be:
```
requires_gnutls_tls1_3                                                          
requires_gnutls_record_size_limit                                               
requires_all_configs_enabled MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C
requires_config_enabled MBEDTLS_SSL_RECORD_SIZE_LIMIT                           
requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
```

ronald-cron-arm reviewed (changes requested):
I've re-reviewed the library code and reviewed the tests.

ronald-cron-arm reviewed (changes requested):
One comment not addressed yet it seems, otherwise this looks good to me. Otherwise, as you have experienced with the tests, currently we can not reach the maximum payload of 16384 in TLS 1.3 records. I have created #8671 to track that.

ronald-cron-arm commented on library/ssl_misc.h:
```diff
@@ -439,6 +439,24 @@ size_t mbedtls_ssl_get_output_max_frag_len(const mbedtls_ssl_context *ssl);
 
+#if defined(MBEDTLS_SSL_RECORD_SIZE_LIMIT)
+/**
+ * \brief          Return the RecordSizeLimit (in bytes) for
+ *                 the output buffer. This is less than the value requested by the
+ *                 peer (see RFC 8449), since it subtracts the space required for the
+ *                 content type and padding of the TLSInnerPlaintext struct (RFC 8446).
+ *                 Returns MBEDTLS_SSL_OUT_CONTENT_LEN if no limit was requested by the peer.
+ *
+ * \sa             mbedtls_ssl_get_max_out_record_payload()
+ *                 ssl_compute_internal_record_size_limit()
+ *
+ * \param ssl      SSL context
+ *
+ * \return         Current record size limit for the output buffer.
```
This is outdated. Rather something like:
```suggestion
 * \brief    Get the size limit in bytes for the protected outgoing records
 *           as defined in RFC 8449
 *
 * \param ssl      SSL context
 *
 * \return         The size limit in bytes for the protected outgoing
 *                 records as defined in RFC 8449.
```

ronald-cron-arm reviewed (changes requested):
The description of mbedtls_ssl_get_output_record_size_limit() is outdated. Otherwise this looks good to me.

ronald-cron-arm reviewed (approved):
LGTM, thanks.

tom-cosgrove-arm commented on library/ssl_tls.c:
```diff
@@ -2610,6 +2618,13 @@ static int ssl_tls13_session_load(mbedtls_ssl_session *session,
 #endif
+#if defined(MBEDTLS_SSL_RECORD_SIZE_LIMIT)
+    if (end - p < 2) {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
```
I'd like to think that if someone was using Mbed TLS just before this PR, and had a saved session, then recompiled with this PR, and tried to resume, they'd get an `MBEDTLS_ERR_SSL_VERSION_MISMATCH` (considering it the version of the data, rather than just the version of the library) error, not a `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` error, although I suspect the caller would treat both identically.

That would suggest that `MBEDTLS_SSL_RECORD_SIZE_LIMIT` should be incorporated in `SSL_SERIALIZED_SESSION_CONFIG_BITFLAG`

However, the same reasoning applies to `MBEDTLS_SSL_EARLY_DATA`, so this looks like pre-existing, rather than anything else, and I'm mainly making the observation here, rather than asking for a change.

ronald-cron-arm replied:
> That would suggest that `MBEDTLS_SSL_RECORD_SIZE_LIMIT` should be incorporated in `SSL_SERIALIZED_SESSION_CONFIG_BITFLAG`

That's a good point. I think we overlooked this aspect of serialized sessions while working on new session tickets, early data and now record size limit. I am creating an issue to address that: #8686.

tom-cosgrove-arm commented on library/ssl_tls.c:
```diff
@@ -3372,6 +3387,31 @@ const char *mbedtls_ssl_get_version(const mbedtls_ssl_context *ssl)
+    size_t record_size_limit = max_len;
+
+    if (ssl->session != NULL &&
+        ssl->session->record_size_limit >= MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN &&
```
It's not clear to me under what circumstances we would put a value < `MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN` into the session. Would it make more sense to verify when populating `ssl->session->record_size_limit` and then be able to skip the check here?

ronald-cron-arm replied:
If we do not receive any record size limit extension then `ssl->session->record_size_limit` is equal to zero. Thus to me this check is to detect if the peer sent a record size limit or not.

tom-cosgrove-arm replied:
Ah - if this test had been for `> 0` then I might have thought that.

Would (in the future, not this PR) it make more sense to set this to the value to use for record size and use it unconditionally?

ronald-cron-arm replied:
Yes that would be an improvement.

tom-cosgrove-arm commented on tests/src/test_helpers/ssl_helpers.c:
```diff
@@ -1776,6 +1776,10 @@ int mbedtls_test_ssl_tls13_populate_session(mbedtls_ssl_session *session,
 #endif /* MBEDTLS_SSL_CLI_C */
 
+#if defined(MBEDTLS_SSL_RECORD_SIZE_LIMIT)
+    session->record_size_limit = 2048;
```
What if `MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN` > 2048?

ronald-cron-arm replied:
MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN is defined by RFC 8449 to be equal to 64. We should probably add a comment where MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN is defined to make that clear.

tom-cosgrove-arm commented on tests/ssl-opt.sh:
```diff
@@ -4837,34 +4837,177 @@ run_test    "Max fragment length: DTLS client, larger message" \
-#            -s "Received alert \[110]: An unsupported extension was sent"
+#            -c "RecordSizeLimit: 16385 Bytes"
+
+# In the following (9) tests, --recordsize is the value used by the G_NEXT_CLI (3.7.2) to configure the
```
(minor; only if making other changes) `9` seems to be a number that won't be maintained (as tests are added and removed in this section, it's likely that at some point we'll forget to update this number) so best removed. Would it be better to have an "end of record size limit tests" comment?

tom-cosgrove-arm commented on tests/ssl-opt.sh:
```diff
@@ -4837,34 +4837,177 @@ run_test    "Max fragment length: DTLS client, larger message" \
+
+# In the following (9) tests, --recordsize is the value used by the G_NEXT_CLI (3.7.2) to configure the
+# maximum record size using "https://gnutls.org/reference/gnutls-gnutls.html#gnutls-record-set-max-size".
+# There is currently a lower limit of 512, caused by this function not respecting the
```
(minor; not sure if worth an update to the PR on its own)

On first (and second!) reading, it wasn't clear to me that "this function" referred to the function documented in the URL on the line above, mainly because it wasn't clear that the object of "using" was a function.

I wonder if it would be better to be explicit about `gnutls_record_set_max_size()` rather than saying "this function", or if "that function" would work just as well.

tom-cosgrove-arm commented on tests/ssl-opt.sh:
```diff
@@ -4837,34 +4837,177 @@ run_test    "Max fragment length: DTLS client, larger message" \
+# https://gnutls.org/reference/gnutls-gnutls.html#gnutls-record-set-max-recv-size.
+# There is currently an upper limit of 4096, caused by the cli arg parser:
+# https://gitlab.com/gnutls/gnutls/-/blob/3.7.2/src/cli-args.def#L395.
+# Thus, these tests are currently limit to that value range.
```
(minor) Best be explicit. Also, should be `limited`
```suggestion
# Thus, these tests are currently limited to the value range 512-4096.
```

tom-cosgrove-arm commented on tests/ssl-opt.sh:
```diff
@@ -4837,34 +4837,177 @@ run_test    "Max fragment length: DTLS client, larger message" \
+# Thus, these tests are currently limit to that value range.
+# Moreover, the value sent in the extension is expected to be larger by one compared
+# to the value passed on the cli:
+# https://gitlab.com/gnutls/gnutls/-/blob/3.7.2/lib/ext/record_size_limit.c#L142
```
Is it `is expected to be` or just `will be`?

`Also, the value sent in the extension will be one larger than the value set at the command line:`?

tom-cosgrove-arm commented on library/ssl_tls.c:
```diff
@@ -3470,6 +3511,30 @@ int mbedtls_ssl_get_max_out_record_payload(const mbedtls_ssl_context *ssl)
+
+    if (ssl->transform_out != NULL &&
+        ssl->transform_out->tls_version == MBEDTLS_SSL_VERSION_TLS1_3) {
+        /* RFC 8449, section 4:
```
It's not immediately clear why this is not within the `MBEDTLS_SSL_RECORD_SIZE_LIMIT` block.  Some of the text in [Ronald's comment](https://github.com/Mbed-TLS/mbedtls/pull/7455#discussion_r1433904997) might help here

tom-cosgrove-arm reviewed (approved):
Looks good to me; nice work.

Some minor comments; I'm not sure if they need to be addressed as part of this PR, or could be handled (if necessary) in the other PR

tom-cosgrove-arm commented:
Do we still need to be able to craft a small certificate for future testing, as mentioned in [Jano' comment](https://github.com/Mbed-TLS/mbedtls/pull/7455#issuecomment-1827635587)?

If so, [removing curves smaller than 255 bits - issue 8136](https://github.com/Mbed-TLS/mbedtls/issues/8136) might be a problem!

ronald-cron-arm commented:
> Do we still need to be able to craft a small certificate for future testing, as mentioned in [Janos' comment](https://github.com/Mbed-TLS/mbedtls/pull/7455#issuecomment-1827635587)?
> 
> If so, [removing curves smaller than 255 bits (issue 8136)](https://github.com/Mbed-TLS/mbedtls/issues/8136) might be a problem!

I do not think that this would improve much the testing of the record size limit code as now we have tests with a record size limit of 512 using a pre-shared key.
